core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -959,7 +965,9 @@ private[spark] class TaskSchedulerImpl(
         barrierCoordinator.stop()
       }
     }
+    starvationFutures.asScala.foreach(_.cancel(false))
```

Comment:
My last concern is, do these Futures build up and cause a potential memory leak?  I don't think they hold onto large result objects. The Future itself isn't big either. So maybe not an issue. I don't know of a clean way to delete them on completion of individual tasks.

Reply:
Good thinking. Let me revert this part. Maybe `ThreadUtils.shutdown(starvationTimer)` is good enough.